keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java

```diff
@@ -18,44 +20,61 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
     }
 
-    public T sample(long[] shape, KeanuRandom random) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        double p = random.nextDouble();
-        double sum = 0;
-
-        if (sumOfProbabilities == 0.0) {
+    public Tensor<T> sample(long[] shape, KeanuRandom random) {
+        DoubleTensor sumOfProbabilities = getSumOfProbabilities(shape);
+        if (containsNonPositiveEntry(sumOfProbabilities)) {
             throw new IllegalArgumentException("Cannot sample from a zero probability setup.");
         }
 
-        T value = null;
+        double p = random.nextDouble();
+        DoubleTensor sum = DoubleTensor.zeros(shape);
+
+        Tensor<T> sample = Tensor.create(null, shape);
+        BooleanTensor sampleValuesSetSoFar = BooleanTensor.falses(shape);
+
         for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
-            sum += entry.getValue().scalar() / sumOfProbabilities;
-            if (p < sum) {
-                value = entry.getKey();
+            sum.plusInPlace(entry.getValue().div(sumOfProbabilities));
+
+            BooleanTensor maskForUnassignedSampleValues = sampleValuesSetSoFar.xor(sum.greaterThan(p));
+            sample = maskForUnassignedSampleValues.where(Tensor.scalar(entry.getKey()), sample);
+
+            sampleValuesSetSoFar.orInPlace(maskForUnassignedSampleValues);
+
+            if (sampleValuesSetSoFar.allTrue()) {
                 break;
             }
         }
-        if (value == null) {
+
+        if (!sampleValuesSetSoFar.allTrue()) {
             T[] values = (T[]) selectableValues.keySet().toArray();
-            value = values[values.length - 1];
+            sample = Tensor.create(values[values.length - 1], shape);
```

Comment:
You can do it without the cast if you avoid raw arrays:

            Set<T> values = selectableValues.keySet();
            T lastValue = ImmutableList.copyOf(values).get(values.size() - 1);
            sample = Tensor.create(lastValue, shape);

Reply:
Cool!